Hack to fix some problematic queries

gortiz opened the pull request:
Queries like:
```sql
WITH CTE_B AS (
  SELECT 1 AS __ts FROM a GROUP BY __ts
) 
SELECT 1 
FROM CTE_B 
WHERE __ts >= 1 AND __ts < 2
```

Fail when they are executed. The reason is that very tricky, but mostly is due to:
- Our rules are not able to get rid of the tautology (`1 >= 1 and 1 < 2` is always true), so it is pushed into the leaf stage
- The leaf stage generates an equivalent V1 query, but does not apply `CompileTimeFunctionsInvoker`, so these tautologies are not removed from the expression tree
- V1 row-by-row engine doesn't know how to execute these tautologies because it only expects expressions like `col1 > doubleLiteral`.

These tautologies are generated in `PinotFilterExpandSearchRule`. We have two rules that should ideally remove them:
- `PinotEvaluateLiteralRule` (which evaluates the expressions using V1 semantics)
- `CoreRules.FILTER_REDUCE_EXPRESSIONS` (which evalautes the expressions using Calcite semantics)

Alternatively, we should be able to call `CompileTimeFunctionsInvoker` rewriter in the leaf stage. This invoker is also using V1 semantics.

But any of these changes works.
1. `PinotEvaluateLiteralRule` and `CompileTimeFunctionsInvoker` partially work because V1 can execute `=` assuming each operand is a number. Things like `1=1` or `'123'='123'` but cannot execute `'a' = 'a'`.
2. `CoreRules.FILTER_REDUCE_EXPRESSIONS` can get rid of the tautologies, but in some scenarios it creates just another `SEARCH` operation. V1 cannot execute them.

I'm not proud of the solution found in this PR. It basically consist of applying the pair `PinotFilterExpandSearchRule`, `CoreRules.FILTER_REDUCE_EXPRESSIONS` twice.  Although it seems to be working in all cases I tried, I'm sure there will be cases were this solution is not good enough.

IMO it would be better to:
1. Have a centralized function register, as suggested by @walterddr long ago.
2. Support `search` in V1.

Do you have other ideas on how to fix this problem?

gortiz committed:
Hack to fix some problematic queries

gortiz commented on pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotFilterExpandSearchRule.java:
```diff
@@ -28,6 +28,18 @@
 import org.apache.calcite.tools.RelBuilderFactory;
 
 
+/**
+ * Rule to expand search condition in filter.
+ * <p>
+ * For example, the filter condition:
+ * <p>
+ *   <code>SEARCH(col1, Sarg[[1..2)])</code>
+ * </p>
+ * Is expanded to:
+ * <p>
+ *   <code>AND(>=(col1, 1), <(col1, 2))</code>
+ *
+ */
```
Boy Scout Rule

gortiz commented on pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java:
```diff
@@ -81,6 +83,7 @@
  * <p>It provide the higher level entry interface to convert a SQL string into a {@link DispatchableSubPlan}.
  */
 public class QueryEnvironment {
+  private static final Logger LOGGER = LoggerFactory.getLogger(QueryEnvironment.class);
```
I've added a bunch of logs to better understand whether the tautologies were removed by the validator or the rules. In the cases I mention it is clear it is always the rules, but it was useful to detect other cases where I actually include the tautology in the SQL request

gortiz replied:
All logs are in debug mode, so they shouldn't be enabled by default.

gortiz commented on pinot-query-planner/src/test/resources/log4j2.xml:
```diff
@@ -0,0 +1,45 @@
+    </Root>
+    <AsyncLogger name="org.reflections" level="error" additivity="false"/>
+    <Logger name="org.apache.calcite.plan.RelOptPlanner" level="debug" additivity="false">
+<!--      Change FULL_PLAN marker from onMatch="DENY" to onMatch='ACCEPT" to see the full plan before and after each
+          rule that modifies the plan is applied -->
+<!--      <MarkerFilter marker="FULL_PLAN" onMatch="ACCEPT" onMismatch="NEUTRAL"/>-->
+      <MarkerFilter marker="FULL_PLAN" onMatch="DENY" onMismatch="NEUTRAL"/>
+      <AppenderRef ref="console"/>
```
I found this very useful to actually see what was going on in the rules. ACCEPT is better, but very verbose. I think it could be useful to have in our CI tests, but I will use DENY for a while to be sure we don't spam too much.

BTW, with this log I discovered that `PinotRelDistributionTraitRule` is being applied multiple times, which may introduce some extra latency in very large plans.

gortiz commented on pinot-query-planner/src/test/resources/queries/LiteralEvaluationPlans.json:
```diff
@@ -259,5 +259,49 @@
+    "queries": [
+      {
+        "description": "Simple filter on constants is simplified",
+        "sql": "EXPLAIN PLAN FOR WITH CTE_B AS (SELECT 1692057600000 AS __ts FROM a GROUP BY __ts) SELECT 1692057600000 AS __ts FROM CTE_B WHERE __ts >= 1692057600000 GROUP BY __ts",
+        "output": [
+          "Execution Plan",
+          "\nLogicalProject(__ts=[1692057600000:BIGINT])",
+          "\n  LogicalAggregate(group=[{0}])",
+          "\n    PinotLogicalExchange(distribution=[hash[0]])",
+          "\n      LogicalAggregate(group=[{0}])",
+          "\n        LogicalProject(__ts=[1692057600000:BIGINT])",
+          "\n          LogicalTableScan(table=[[default, a]])",
+          "\n"
```
This query works in master. Our `PinotEvaluateLiteralRule.Filter.INSTANCE` is able to get rid of it.

yashmayya replied:
This should no longer be an issue (https://github.com/apache/pinot/pull/13711).

gortiz commented on pinot-query-planner/src/test/resources/queries/LiteralEvaluationPlans.json:
```diff
@@ -259,5 +259,49 @@
+      },
+      {
+        "description": "And filter on constants is simplified",
+        "sql": "EXPLAIN PLAN FOR WITH CTE_B AS (SELECT 1692057600000 AS __ts FROM a GROUP BY __ts) SELECT 1692057600000 AS __ts FROM CTE_B WHERE __ts >= 1692057600000 AND __ts < 1693267200000 GROUP BY __ts",
+        "output": [
+          "Execution Plan",
+          "\nLogicalProject(__ts=[1692057600000:BIGINT])",
+          "\n  LogicalAggregate(group=[{0}])",
+          "\n    PinotLogicalExchange(distribution=[hash[0]])",
+          "\n      LogicalAggregate(group=[{0}])",
+          "\n        LogicalProject(__ts=[1692057600000:BIGINT])",
+          "\n          LogicalTableScan(table=[[default, a]])",
+          "\n"
```
This is similar to the query in the PR description. It can be fixed by applying `PinotEvaluateLiteralRule` after `PinotFilterExpandSearchRule` or adding `CompileTimeFunctionsInvoker` into `ServerPlanRequestUtils.QUERY_REWRITERS`.

gortiz commented on pinot-query-planner/src/test/resources/queries/LiteralEvaluationPlans.json:
```diff
@@ -259,5 +259,49 @@
+        ]
+      },
+      {
+        "description": "Search + OR filter on constants is simplified",
+        "sql": "EXPLAIN PLAN FOR WITH tmp2 AS (SELECT CASE WHEN col2 = 'VAL1' THEN 'A' ELSE col2 END AS cased FROM a) SELECT 1 FROM tmp2 WHERE ((cased = 'B') OR (cased = 'A'))",
+        "output": [
+          "Execution Plan",
+          "\nLogicalProject(EXPR$0=[1])",
+          "\n  LogicalFilter(condition=[SEARCH($1, Sarg[_UTF-8'A':VARCHAR CHARACTER SET \"UTF-8\", _UTF-8'B':VARCHAR CHARACTER SET \"UTF-8\", _UTF-8'VAL1':VARCHAR CHARACTER SET \"UTF-8\"]:VARCHAR CHARACTER SET \"UTF-8\")])",
+          "\n    LogicalTableScan(table=[[default, a]])",
+          "\n"
```
This query fails in master and *cannot* be solved by applying `PinotEvaluateLiteralRule` after `PinotFilterExpandSearchRule` or adding `CompileTimeFunctionsInvoker` into `ServerPlanRequestUtils.QUERY_REWRITERS`.

But it works if we apply `CoreRules.FILTER_REDUCE_EXPRESSIONS` followed by `PinotFilterExpandSearchRule` twice in a row.

gortiz committed:
Remove RelOptPlanner logs in tests

It is very verbose

gortiz committed:
Fix one of the new queries, which expected the old (and incorrect) plan

gortiz commented:
Closing this PR as superseded.

This was always a self-admitted hack, and the proper fixes I suggested in the description — *"support SEARCH in V1"* / centralized handling instead of expanding SEARCH into filter trees — were implemented upstream by #14615 ("Improve performance of multi-stage queries with large IN clauses", merged Mar 2025).

That PR:
- Removed `PinotFilterExpandSearchRule` entirely, so SEARCH nodes are no longer expanded into AND/OR predicate trees at planning time.
- Handles SEARCH directly in `RexExpressionUtils`, and — most relevant here — when the left operand is a literal, evaluates the `Sarg` to `Literal.TRUE`/`Literal.FALSE` at conversion time (`evaluateLiteralIn` / `evaluateLiteralNotIn` / `evaluateLiteralOrRanges`). This eliminates the tautology at its source rather than relying on a second `FILTER_REDUCE_EXPRESSIONS` pass to clean it up.
- Added regression coverage (`literal_planning_cte_tests` in `LiteralEvaluationPlans.json`) for the exact CTE-with-literal-filter scenario this PR targeted; it now plans to `LogicalValues(tuples=[[]])`.

So the root cause is gone and the rule this PR modified no longer exists. The only independently-useful leftovers were the per-stage debug logging in `QueryEnvironment` and the test `log4j2.xml`; if still wanted, those can be picked up in a small standalone PR.
